File: sdks/mobile/ios/ZeekySDK/Sources/ZeekySDK/ZeekySDK.swift
import Foundation
import Alamofire
import SwiftyJSON

/// Main Zeeky SDK class for iOS
public class ZeekySDK {
    
    // MARK: - Properties
    
    private let apiKey: String
    private let baseURL: String
    private let session: Session
    private var analyticsEnabled: Bool = false
    private var userProperties: [String: String] = [:]
    
    // MARK: - Initialization
    
    /// Initialize the Zeeky SDK
    /// - Parameters:
    ///   - apiKey: Your Zeeky API key
    ///   - baseURL: Base URL for the Zeeky API (optional, defaults to production)
    public init(apiKey: String, baseURL: String = "https://api.zeeky.ai") {
        self.apiKey = apiKey
        self.baseURL = baseURL
        
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 60
        
        self.session = Session(configuration: configuration)
    }
    
    // MARK: - Core Methods
    
    /// Send a text message to Zeeky
    /// - Parameters:
    ///   - text: The message text
    ///   - completion: Completion handler with result
    public func sendMessage(_ text: String, completion: @escaping (Result<ZeekyResponse, ZeekyError>) -> Void) {
        let parameters: [String: Any] = [
            "message": text,
            "timestamp": ISO8601DateFormatter().string(from: Date())
        ]
        
        let headers: HTTPHeaders = [
            "Authorization": "Bearer \(apiKey)",
            "Content-Type": "application/json"
        ]
        
        session.request(
            "\(baseURL)/v1/messages",
            method: .post,
            parameters: parameters,
            encoding: JSONEncoding.default,
            headers: headers
        ).validate().responseData { [weak self] response in
            switch response.result {
            case .success(let data):
                do {
                    let json = try JSON(data: data)
                    let zeekyResponse = try self?.parseResponse(json) ?? ZeekyResponse(text: "Error parsing response")
                    completion(.success(zeekyResponse))
                    
                    if self?.analyticsEnabled == true {
                        self?.trackEvent("message_sent", properties: ["message_length": text.count])
                    }
                } catch {
                    completion(.failure(.parsingError))
                }
            case .failure(let error):
                completion(.failure(.networkError(error)))
            }
        }
    }
    
    /// Get available plugins
    /// - Parameter completion: Completion handler with result
    public func getAvailablePlugins(completion: @escaping (Result<[ZeekyPlugin], ZeekyError>) -> Void) {
        let headers: HTTPHeaders = [
            "Authorization": "Bearer \(apiKey)"
        ]
        
        session.request(
            "\(baseURL)/v1/plugins",
            method: .get,
            headers: headers
        ).validate().responseData { response in
            switch response.result {
            case .success(let data):
                do {
                    let json = try JSON(data: data)
                    let plugins = try self.parsePlugins(json)
                    completion(.success(plugins))
                } catch {
                    completion(.failure(.parsingError))
                }
            case .failure(let error):
                completion(.failure(.networkError(error)))
            }
        }
    }
    
    /// Execute a specific plugin
    /// - Parameters:
    ///   - pluginId: The plugin ID to execute
    ///   - parameters: Parameters for the plugin
    ///   - completion: Completion handler with result
    public func executePlugin(_ pluginId: String, parameters: [String: Any], completion: @escaping (Result<ZeekyResponse, ZeekyError>) -> Void) {
        let requestBody: [String: Any] = [
            "plugin_id": pluginId,
            "parameters": parameters,
            "timestamp": ISO8601DateFormatter().string(from: Date())
        ]
        
        let headers: HTTPHeaders = [
            "Authorization": "Bearer \(apiKey)",
            "Content-Type": "application/json"
        ]
        
        session.request(
            "\(baseURL)/v1/plugins/execute",
            method: .post,
            parameters: requestBody,
            encoding: JSONEncoding.default,
            headers: headers
        ).validate().responseData { response in
            switch response.result {
            case .success(let data):
                do {
                    let json = try JSON(data: data)
                    let zeekyResponse = try self.parseResponse(json)
                    completion(.success(zeekyResponse))
                } catch {
                    completion(.failure(.parsingError))
                }
            case .failure(let error):
                completion(.failure(.networkError(error)))
            }
        }
    }
    
    // MARK: - Voice Interface
    
    /// Enable voice input (placeholder for future implementation)
    /// - Parameter completion: Completion handler with transcribed text
    public func enableVoiceInput(completion: @escaping (String) -> Void) {
        // TODO: Implement speech recognition
        completion("Voice input not yet implemented")
    }
    
    /// Synthesize speech from text (placeholder for future implementation)
    /// - Parameters:
    ///   - text: Text to synthesize
    ///   - completion: Completion handler with audio data
    public func synthesizeSpeech(_ text: String, completion: @escaping (Result<Data, ZeekyError>) -> Void) {
        // TODO: Implement text-to-speech
        completion(.failure(.featureNotAvailable))
    }
    
    // MARK: - Analytics
    
    /// Enable analytics tracking
    /// - Parameter enabled: Whether to enable analytics
    public func enableAnalytics(_ enabled: Bool) {
        self.analyticsEnabled = enabled
    }
    
    /// Set user properties for analytics
    /// - Parameter properties: User properties dictionary
    public func setUserProperties(_ properties: [String: String]) {
        self.userProperties = properties
    }
    
    /// Track a custom event
    /// - Parameters:
    ///   - eventName: Name of the event
    ///   - properties: Event properties
    private func trackEvent(_ eventName: String, properties: [String: Any] = [:]) {
        guard analyticsEnabled else { return }
        
        let eventData: [String: Any] = [
            "event": eventName,
            "properties": properties,
            "user_properties": userProperties,
            "timestamp": ISO8601DateFormatter().string(from: Date())
        ]
        
        let headers: HTTPHeaders = [
            "Authorization": "Bearer \(apiKey)",
            "Content-Type": "application/json"
        ]
        
        session.request(
            "\(baseURL)/v1/analytics/track",
            method: .post,
            parameters: eventData,
            encoding: JSONEncoding.default,
            headers: headers
        ).response { _ in
            // Analytics tracking is fire-and-forget
        }
    }
    
    // MARK: - Private Methods
    
    private func parseResponse(_ json: JSON) throws -> ZeekyResponse {
        let text = json["response"]["text"].stringValue
        let metadata = json["response"]["metadata"].dictionaryObject ?? [:]
        let timestamp = Date()
        let pluginId = json["response"]["plugin_id"].string
        
        return ZeekyResponse(
            text: text,
            metadata: metadata,
            timestamp: timestamp,
            pluginId: pluginId
        )
    }
    
    private func parsePlugins(_ json: JSON) throws -> [ZeekyPlugin] {
        var plugins: [ZeekyPlugin] = []
        
        for (_, pluginJson) in json["plugins"] {
            let plugin = ZeekyPlugin(
                id: pluginJson["id"].stringValue,
                name: pluginJson["name"].stringValue,
                description: pluginJson["description"].stringValue,
                capabilities: pluginJson["capabilities"].arrayObject as? [String] ?? [],
                version: pluginJson["version"].stringValue
            )
            plugins.append(plugin)
        }
        
        return plugins
    }
}

File: sdks/mobile/android/zeeky-android-sdk/src/main/java/com/zeeky/sdk/ZeekySDK.kt
package com.zeeky.sdk

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.util.Date

/**
 * Main Zeeky SDK class for Android
 */
class ZeekySDK private constructor(
    private val context: Context,
    private val apiKey: String,
    private val baseUrl: String
) {
    
    private val retrofit: Retrofit
    private val apiService: ZeekyApiService
    private var analyticsEnabled: Boolean = false
    private var userProperties: Map<String, String> = emptyMap()
    
    init {
        retrofit = Retrofit.Builder()
            .baseUrl(baseUrl)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
        
        apiService = retrofit.create(ZeekyApiService::class.java)
    }
    
    companion object {
        private const val DEFAULT_BASE_URL = "https://api.zeeky.ai"
        
        fun create(
            context: Context,
            apiKey: String,
            baseUrl: String = DEFAULT_BASE_URL
        ): ZeekySDK {
            return ZeekySDK(context, apiKey, baseUrl)
        }
    }
    
    /**
     * Send a text message to Zeeky
     */
    fun sendMessage(
        text: String,
        callback: (Result<ZeekyResponse>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val request = MessageRequest(
                    message = text,
                    timestamp = Date().toString()
                )
                
                val response = apiService.sendMessage(
                    authorization = "Bearer $apiKey",
                    request = request
                )
                
                withContext(Dispatchers.Main) {
                    callback(Result.success(response))
                }
                
                if (analyticsEnabled) {
                    trackEvent("message_sent", mapOf("message_length" to text.length.toString()))
                }
                
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    callback(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Get available plugins
     */
    fun getAvailablePlugins(
        callback: (Result<List<ZeekyPlugin>>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = apiService.getPlugins(authorization = "Bearer $apiKey")
                
                withContext(Dispatchers.Main) {
                    callback(Result.success(response.plugins))
                }
                
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    callback(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Execute a specific plugin
     */
    fun executePlugin(
        pluginId: String,
        parameters: Map<String, Any>,
        callback: (Result<ZeekyResponse>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val request = PluginExecuteRequest(
                    pluginId = pluginId,
                    parameters = parameters,
                    timestamp = Date().toString()
                )
                
                val response = apiService.executePlugin(
                    authorization = "Bearer $apiKey",
                    request = request
                )
                
                withContext(Dispatchers.Main) {
                    callback(Result.success(response))
                }
                
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    callback(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Enable voice input (placeholder for future implementation)
     */
    fun enableVoiceInput(callback: (String) -> Unit) {
        // TODO: Implement speech recognition
        callback("Voice input not yet implemented")
    }
    
    /**
     * Synthesize speech from text (placeholder for future implementation)
     */
    fun synthesizeSpeech(
        text: String,
        callback: (Result<ByteArray>) -> Unit
    ) {
        // TODO: Implement text-to-speech
        callback(Result.failure(Exception("Feature not yet implemented")))
    }
    
    /**
     * Enable analytics tracking
     */
    fun enableAnalytics(enabled: Boolean) {
        this.analyticsEnabled = enabled
    }
    
    /**
     * Set user properties for analytics
     */
    fun setUserProperties(properties: Map<String, String>) {
        this.userProperties = properties
    }
    
    /**
     * Track a custom event
     */
    private fun trackEvent(eventName: String, properties: Map<String, String> = emptyMap()) {
        if (!analyticsEnabled) return
        
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val eventData = AnalyticsEvent(
                    event = eventName,
                    properties = properties,
                    userProperties = userProperties,
                    timestamp = Date().toString()
                )
                
                apiService.trackEvent(
                    authorization = "Bearer $apiKey",
                    event = eventData
                )
            } catch (e: Exception) {
                // Analytics tracking is fire-and-forget
            }
        }
    }
}

/**
 * API service interface
 */
interface ZeekyApiService {
    
    @POST("v1/messages")
    suspend fun sendMessage(
        @Header("Authorization") authorization: String,
        @Body request: MessageRequest
    ): ZeekyResponse
    
    @GET("v1/plugins")
    suspend fun getPlugins(
        @Header("Authorization") authorization: String
    ): PluginsResponse
    
    @POST("v1/plugins/execute")
    suspend fun executePlugin(
        @Header("Authorization") authorization: String,
        @Body request: PluginExecuteRequest
    ): ZeekyResponse
    
    @POST("v1/analytics/track")
    suspend fun trackEvent(
        @Header("Authorization") authorization: String,
        @Body event: AnalyticsEvent
    )
}

/**
 * Data classes
 */
data class MessageRequest(
    val message: String,
    val timestamp: String
)

data class PluginExecuteRequest(
    val pluginId: String,
    val parameters: Map<String, Any>,
    val timestamp: String
)

data class ZeekyResponse(
    val text: String,
    val metadata: Map<String, Any> = emptyMap(),
    val timestamp: String = Date().toString(),
    val pluginId: String? = null
)

data class ZeekyPlugin(
    val id: String,
    val name: String,
    val description: String,
    val capabilities: List<String>,
    val version: String
)

data class PluginsResponse(
    val plugins: List<ZeekyPlugin>
)

data class AnalyticsEvent(
    val event: String,
    val properties: Map<String, String> = emptyMap(),
    val userProperties: Map<String, String> = emptyMap(),
    val timestamp: String
)